### PR TITLE
revise tilecache

### PR DESCRIPTION
**Pull request by jcupitt**
quite a bit simpler, perhaps faster

see https://github.com/libvips/libvips/issues/3890

### CODE REVIEW

**Comment by jcupitt** [comment]
Needs quite a bit of testing and benchmarking! But at least simple cases seem to work.

**Comment by jcupitt** [comment]
This PR makes threads that request tiles compute downstream tiles cooperatively. It should (I hope) handle overlapping requests.

Threads loop every tile to see if their set of pending tiles have been computed, we could get rid of this, though I doubt if it's a bottleneck. But who knows!

Other benchmarks would be useful for testing.

**Comment by jcupitt** [comment]
I've not found a case where it's faster yet :( In fact overall it's slightly slower.

I'll get the test suite to pass, perhaps I missed something.

**Comment by kleisauke** [comment]
Either removing `tile-width=256` here:
https://github.com/libvips/libvips/blob/566ce6d470faaa70b6ff9c7d02bbbc219576e2be/test/test-suite/test_foreign.py#L592-L593

Or disabling threaded access here:
https://github.com/libvips/libvips/blob/566ce6d470faaa70b6ff9c7d02bbbc219576e2be/libvips/foreign/tiff2vips.c#L2614

Makes the test suite pass, fwiw.

File: libvips/conversion/tilecache.c [hunk 1]
```diff
@@ -602,156 +670,59 @@ static int
...
-				break;
-			}
+	/* Loop until all our tiles are ready, or there's an error of some sort,
+	 * perhaps a loader hits a bad tile.
+	 */
+	GSList *todo;
+	todo = g_slist_copy(work);
+	while (todo) {
+		VipsTile *tile = (VipsTile *) todo->data;
+		todo = g_slist_remove(todo, tile);
```
Comment (kleisauke) [inline libvips/conversion/tilecache.c:704]: I think we could avoid the (possible expensive) `g_slist_copy()` here.

```suggestion
	for (GSList *todo = work; todo; todo = todo->next) {
		VipsTile *tile = (VipsTile *) todo->data;
```

File: libvips/conversion/tilecache.c [hunk 2]
```diff
@@ -602,156 +670,59 @@ static int
...
+
+		if (tile->dirty &&
+			vips_tile_cache_compute(in_region, tile)) {
+			*stop = TRUE;
+			break;
 		}
+	}
+	VIPS_FREEF(g_slist_free, todo);
```
Comment (kleisauke) [inline libvips/conversion/tilecache.c:712]: `g_slist_free()` is not needed if the other suggested change is accepted.

```suggestion
```

**Pull request closed without merging**